src/utils/logging.jl

```diff
@@ -0,0 +1,276 @@
+    if _is_level_valid(tracker, event.level)
+        if haskey(tracker.events[event.level], event.id)
+            tracker.events[event.level][event.id].count += 1
+            if suppressed
+                tracker.events[event.level][event.id].suppressed += 1
+            end
+        else
+            tracker.events[event.level][event.id] = event
+        end
+    end
+end
+
+function _is_level_valid(tracker::LogEventTracker, level::Logging.LogLevel)
+    return level in keys(tracker.events)
+end
+
+"""
+    MultiLogger(Array{AbstractLogger}, Union{LogEventTracker, Nothing})
+
+Redirects log events to multiple loggers. The primary use case is to allow logging to
+both a file and the console. Secondarily, it can track the counts of all log messages.
+
+# Example
+```julia
+MultiLogger([ConsoleLogger(stderr), SimpleLogger(stream)], LogEventTracker())
+```
+"""
+mutable struct MultiLogger <: Logging.AbstractLogger
+    loggers::Array{Logging.AbstractLogger}
+    tracker::Union{LogEventTracker, Nothing}
+end
+
+"""
+Creates a MultiLogger with no event tracking.
+
+# Example
+```julia
+MultiLogger([ConsoleLogger(stderr), SimpleLogger(stream)])
+```
+"""
+function MultiLogger(loggers::Array{T}) where T <: Logging.AbstractLogger
+    return MultiLogger(loggers, nothing)
+end
+
+Logging.shouldlog(logger::MultiLogger, level, _module, group, id) = true
+
+function Logging.min_enabled_level(logger::MultiLogger)
+    return minimum([Logging.min_enabled_level(x) for x in logger.loggers])
+end
+
+Logging.catch_exceptions(logger::MultiLogger) = false
+
+function Logging.handle_message(logger::MultiLogger,
+                                level,
+                                message,
+                                _module,
+                                group,
+                                id,
+                                file,
+                                line;
+                                maxlog=nothing,
+                                kwargs...)
+    suppressed = false
+    for logger_ in logger.loggers
+        if level >= Logging.min_enabled_level(logger_)
+            if Logging.shouldlog(logger_, level, _module, group, id)
+                Logging.handle_message(logger_, level, message, _module, group, id, file,
+                                       line; maxlog=maxlog, kwargs...)
+            else
+                suppressed = true
+            end
+        end
+    end
+
+    if logger.tracker != nothing
+        event = LogEvent(file, line, id, message, level)
+        increment_count(logger.tracker, event, suppressed)
+    end
+
+    return
+end
+
+"""Returns a summary of log event counts by level."""
+function report_log_summary(logger::MultiLogger)::String
+    if logger.tracker == nothing
+        error("log event tracking is not enabled")
+    end
+
+    return report_log_summary(logger.tracker)
+end
+
+"""Flush any file streams."""
+function Base.flush(logger::MultiLogger)
```

Comment:
Note that I added this after our discussion today.  This is what Julia devs have frequently posted on forums when people ask why log files haven't been updated.